In inductor triton generated code, avoid masking when numel=1

fdrocha opened the pull request:
This is implementing an idea from @lezcano : if we have a generated triton kernel with `xnumel=1`, then `xmask` is just `0<1` and can be dropped from all `load`/`store`/`where`.

The `xnumel=1` case actually comes up relatively often when code for reductions is being generated. @lezcano reported some  performance gains in micro-benchmarks (see comment below) and it is a very simple change.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91254



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Commit:
In inductor triton generated code, avoid masking when numel=1

[ghstack-poisoned]

lezcano commented:
For context `xnumel = 1` appears in the case of contiguous reductions. The gains of this optimisation were around 2-3% on a reduction of size `(1024, 64k)` on an A100. And, as a byproduct, we get simpler generated code, which is nice :)

jansel reviewed (changes requested):
Can you add a test for:
1) A single element pointwise (not reduction)
2) An indirect single element pointwise

ngimel commented:
There is a more general optimization here for `xnumel` that is a multiple of XBLOCK?

Commit:
Update on "In inductor triton generated code, avoid masking when numel=1"

This is implementing an idea from lezcano : if we have a generated triton kernel with `xnumel=1`, then `xmask` is just `0<1` and can be dropped from all `load`/`store`/`where`.

The `xnumel=1` case actually comes up relatively often when code for reductions is being generated. lezcano reported some  performance gains in micro-benchmarks (see comment below) and it is a very simple change.





cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

fdrocha commented:
> Can you add a test for:
> 
> 1. A single element pointwise (not reduction)
> 2. An indirect single element pointwise

I added a couple of tests in second commit. Not sure if it is what you had in mind, could you take a look and let me know?

ngimel commented:
In this PR, you are relying on XBLOCK being 1 if xnumel is 1, which is, strictly speaking, not guaranteed (it's true today, but there are no hard guarantees around it), would be good to at least leave comments in related places in autotune that set XBLOCK to that effect. 
As for guarantees on XBLOCK, triton requires XBLOCK to be power of 2, and we don't expect XBLOCK to be larger than 1024 (we are playing around with idea of setting RBLOCK to larger values when needed).

Commit:
Update on "In inductor triton generated code, avoid masking when numel=1"

This is implementing an idea from lezcano : if we have a generated triton kernel with `xnumel=1`, then `xmask` is just `0<1` and can be dropped from all `load`/`store`/`where`.

The `xnumel=1` case actually comes up relatively often when code for reductions is being generated. lezcano reported some  performance gains in micro-benchmarks (see comment below) and it is a very simple change.





cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

fdrocha commented:
> In this PR, you are relying on XBLOCK being 1 if xnumel is 1, which is, strictly speaking, not guaranteed (it's true today, but there are no hard guarantees around it), would be good to at least leave comments in related places in autotune that set XBLOCK to that effect. As for guarantees on XBLOCK, triton requires XBLOCK to be power of 2, and we don't expect XBLOCK to be larger than 1024 (we are playing around with idea of setting RBLOCK to larger values when needed).

Thanks. I added logic that will drop `xmask` if `xnumel` is a multiple of 1024.
I also added some asserts in `autotune.grid` to make sure that both conditions hold (`xnumel=1 => XBLOCK=1` and `XBLOCK` divides 1024). PTAL

fdrocha commented:
@ngimel, could you PTAL? Are you okay with last commit?

ngimel commented on torch/_inductor/triton_ops/autotune.py:
```diff
@@ -675,29 +675,47 @@ def mm_autotune(get_io_bound_configs=False):
+    def get_grid_dim(numel, block_name, meta):
+        if numel is None:
+            return None
+        block = meta[block_name]
```
nit: you don't need all of `meta`, you only need `meta[*BLOCK]` argument to this function

fdrocha replied:
Sure, changed that.

ngimel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -782,10 +783,15 @@ def indexing(
+            if tree.prefix.upper() not in config.triton.max_block:
+                continue
+            max_block = config.triton.max_block[tree.prefix.upper()]
+            if tree.numel == 1 or tree.numel % max_block == 0:
```
cc @chillee, this modulo will probably never be true with symbolic shapes? To enable this mask clearing we'd need to look at size hints and set up corresponding guards?

Chillee replied:
Yeah, that sounds right.

ngimel reviewed (approved):
Yeah this looks good, I left a minor comment

Commit:
Update on "In inductor triton generated code, avoid masking when numel=1"

This is implementing an idea from lezcano : if we have a generated triton kernel with `xnumel=1`, then `xmask` is just `0<1` and can be dropped from all `load`/`store`/`where`.

The `xnumel=1` case actually comes up relatively often when code for reductions is being generated. lezcano reported some  performance gains in micro-benchmarks (see comment below) and it is a very simple change.





cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

fdrocha commented:
@pytorchbot merge -g

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -3731,6 +3731,45 @@ def fn(in_ptr0, in_ptr1, in_ptr2):
+            c = a[b] + 1
+            return (c,)
+
+        self.common(
+            fn,
+            (
+                torch.randn(
+                    1,
+                ),
+                torch.tensor(
+                    [
+                        0,
+                    ],
+                    dtype=torch.int64,
+                ),
+            ),
+        )
```
flake8 go home, you're drunk.

lezcano commented:
The triton errors seems unrelated, but I'm not 100% sure...

Commit:
Update on "In inductor triton generated code, avoid masking when numel=1"

This is implementing an idea from lezcano : if we have a generated triton kernel with `xnumel=1`, then `xmask` is just `0<1` and can be dropped from all `load`/`store`/`where`.

The `xnumel=1` case actually comes up relatively often when code for reductions is being generated. lezcano reported some  performance gains in micro-benchmarks (see comment below) and it is a very simple change.





cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

fdrocha commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -f "failures are unrelated"

ngimel commented:
Heads up @fdrocha this PR appears to regress TorchBench and HF perf on the dashboard https://github.com/pytorch/torchdynamo/issues/681#issuecomment-1380879549, and for a smoke perf command
`python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --float16 --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt) --only hf_Bert`, so I'm going to revert it for now, can you please look into why that happens? Timm seems to slightly improve, but I didn't look at the per-model numbers.

ngimel commented:
@pytorchbot revert -m "regresses perf of hf models"

ngimel commented:
@pytorchbot revert -m "regresses perf of hf models" -c nosignal

fdrocha commented:
Handled in #92748
